### ISSUE TITLE
level-dimension 

### ISSUE THREAD

**Issue by houph4** [NVlabs/FasterViT#55]
Thanks for your job,I used **Segformer** before in my work.**FasterVit** has a similiar dimension structure with its encoder.When i deal with image size 320*320.
In **Segformer** encoder,The dimension of image change is (1,3,320,320),s1:(1,64,80,80),s2:(1,128,40,40),s3:(1,256,20,20),s4:(1,512,10,10).
In **faster_vit_0_any_res** encoder,I add a print(x.shape) behind Class FasterVit()-(forward_features)-(x=level).
The dimension of image change is (1,3,320,320),s1:(1,128,40,40),s2:(1,256,20,20),s3:(1,512,10,10),s4:(1,512,10,10).
I don't know why s3&s4 have the same dimension and s1 H$W//4 should be (80,80).

And i find in PatchEmbed,the dimension is (1,64,80,80) not H$W//2 
the first ConV-nn.Conv2d(in_chans, in_dim, 3, 2, 1, bias=False) is correct not (3,1,1) in fig3?

**Comment by ahatamiz** [comment]
Hi @niu17 my appologies for a late reply on this issue. 

I have taken these steps to reproduce the issue:

First, I created a `faster_vit_0_any_res` model as in the following:

```
@register_pip_model
@register_model
def faster_vit_0_any_res(pretrained=False, **kwargs):
    depths = kwargs.pop("depths", [2, 3, 6, 5])
    num_heads = kwargs.pop("num_heads", [2, 4, 8, 16])
    window_size = kwargs.pop("window_size", [7, 7, 10, 10])
    ct_size = kwargs.pop("ct_size", 8)
    dim = kwargs.pop("dim", 64)
    in_dim = kwargs.pop("in_dim", 64)
    mlp_ratio = kwargs.pop("mlp_ratio", 4)
    resolution = kwargs.pop("resolution", [320, 320])
    drop_path_rate = kwargs.pop("drop_path_rate", 0.2)
    model_path = kwargs.pop("model_path", "/tmp/faster_vit_0.pth.tar")
    hat = kwargs.pop("hat", [False, False, True, False])
    model = FasterViT(depths=depths,
                      num_heads=num_heads,
                      window_size=window_size,
                      ct_size=ct_size,
                      dim=dim,
                      in_dim=in_dim,
                      mlp_ratio=mlp_ratio,
                      resolution=resolution,
                      drop_path_rate=drop_path_rate,
                      hat=hat,
                      **kwargs)
    model.default_cfg = default_cfgs['faster_vit_0_any_res']
    if pretrained:
        if not Path(model_path).is_file():
            url = model.default_cfg['url']
            torch.hub.download_url_to_file(url=url, dst=model_path)
        model._load_state_dict(model_path)
    return model
```

Note that I adjusted the window_sizes from `[7,7,7,7]` to `[7,7,10,10]` to account for the fact the input resolution is now `320x320`. 

Next, I simply added print statements:

```
def forward_features(self, x):
    import pdb;
    x = self.patch_embed(x)
    print(x.shape)
    for level in self.levels:
        x = level(x)
        print(x.shape)
    pdb.set_trace()
    x = self.norm(x)
    x = self.avgpool(x)
    x = torch.flatten(x, 1)
    return x
```

I assume this is consistent with your observation. Now here, I would like to make the following comments:

> **why output of  PatchEmbed,the dimension is (1,64,80,80) not H$W//2**

This is expected as the `PatchEmbed`, in ImageNet-pretrained networks, reduce the input resolution by a factor of `4`. For example, `224x224` > `[56,28,14,7]`. In addition to our work, previous networks such as `Swin`, `ConvNext`, etc. also attain the same resolution scheme throughput their architecture. 

Now, if it is desired to also use the output feature map of `PatchEmbed` at `(1,64,160,160)`, please kindly make this modification to `PatchEmbed`:

class PatchEmbed(nn.Module):
    """
    Patch embedding block based on: "Hatamizadeh et al.,
    FasterViT: Fast Vision Transformers with Hierarchical Attention
    """

    def __init__(self, in_chans=3, in_dim=64, dim=96):
        """
        Args:
            in_chans: number of input channels.
            dim: feature size dimension.
        """
        super().__init__()
        self.proj = nn.Identity()
        self.conv_down = nn.Sequential(
            nn.Conv2d(in_chans, in_dim, 3, 2, 1, bias=False),
            nn.BatchNorm2d(in_dim, eps=1e-4),
            nn.ReLU(),
            nn.Conv2d(in_dim, dim, 3, 2, 1, bias=False),
            nn.BatchNorm2d(dim, eps=1e-4),
            nn.ReLU()
            )

    def forward(self, x):
        x = self.proj(x)
        x_out = self.conv_down[0:-3](x) # torch.Size([1, 64, 160, 160])
        x = self.conv_down[-3::](x_out) # torch.Size([1, 64, 80, 80])
        return x, x_out

Now x_out has the shape of `[1, 64, 160, 160]`. 


> **I don't know why s3&s4 have the same dimension and s1 H$W//4 should be (80,80).**

This is also expected as we pass the downsampled feature map to the next stage, and we don't use downsampling in stage `4`. Hence, the output of feature 3 and feature `4` will end up having the same resolution with above print statement. 

In order to address, this please simply make this modification to output the non downsampled features as well: 

```
def forward(self, x):
        B, C, H, W = x.shape
        if self.transformer_block:
            pad_r = (self.window_size - W % self.window_size) % self.window_size
            pad_b = (self.window_size - H % self.window_size) % self.window_size
            if pad_r > 0 or pad_b > 0:
                x = torch.nn.functional.pad(x, (0,pad_r,0,pad_b))
                _, _, Hp, Wp = x.shape
            else:
                Hp, Wp = H, W
        ct = self.global_tokenizer(x) if self.do_gt else None
        if self.transformer_block:
            x = window_partition(x, self.window_size)
        for bn, blk in enumerate(self.blocks):
            x, ct = blk(x, ct)
        if self.transformer_block:
            x = window_reverse(x, self
                               .window_size, Hp, Wp, B)
            if pad_r > 0 or pad_b > 0:
                x = x[:, :, :H, :W].contiguous()
        if self.downsample is None:
            return x, x
        return self.downsample(x), x
```

And finally, we can extract all feature according to: 

```
def forward_features(self, x):
        x_outputs= []
        x, x_out = self.patch_embed(x)
        x_outputs.append(x_out)
        for level in self.levels:
            x, x_out = level(x)
            x_outputs.append(x_out)
        return x_outputs

    def forward(self, x):
        x_outputs = self.forward_features(x)
        return x
```

Hence x_out can be used along with other extracted feature maps. Here I also removed the classification head `self.head(x)`. It would also be preferred if you could please adjust how these output feature representations need to be utilized based on the specific segmentation network -- there could be some modifications needed. 

> **the first ConV-nn.Conv2d(in_chans, in_dim, 3, 2, 1, bias=False) is correct not (3,1,1) in fig3?**

The Figure 3 is correct. We just apply two consecutive convolutional layers as explained in the above. 

I hope the above addresses your concerns.

**Comment by ahatamiz** [comment]
Hi @niu17 ! 

We have just released the [object detection repository](https://github.com/NVlabs/FasterViT/tree/main/object_detection) for FasterViT. 

The FasterViT backbone with required changes can be found [here](https://github.com/NVlabs/FasterViT/blob/main/object_detection/models/dino/fastervit.py). 

We will add pretrained checkpoints very soon ! 

I hope you find this to be useful.
